### PR TITLE
Pool reth

### PR DESCRIPTION
**Pull request by skellet0r**
### Pool Info

* Name: `rETH`
* Base template: [`ETH-TEMPLATE`](https://github.com/curvefi/curve-contract/blob/master/contracts/pool-templates/eth/SwapTemplateEth.vy)

rETH is a staking derivative that accrues interest by increasing the redemption rate cToken style.  

#### StaFi rETH

**General Info**
* Deployment address: [0x9559Aaa82d9649C7A7b220E7c461d2E74c9a3593](https://etherscan.io/address/0x9559aaa82d9649c7a7b220e7c461d2e74c9a3593)
* Link to the token source code: [Etherscan](https://etherscan.io/address/0x9559aaa82d9649c7a7b220e7c461d2e74c9a3593#code) | [Github Repo](https://github.com/stafiprotocol/eth2-staking)
* Link to the developer documentation: [Dev Docs-ish](https://docs.stafi.io/rproduct/reth-solution)

**Behaviours**
1. On a successful transfer this token: returns True
2. On a failed transfer this token: reverts
3. Are there any special approval behaviors? No special behaviors
4. Is there a transfer fee, or the capability to add a transfer fee in the future? No transfer fee, nor ability to add in the future

- [x] I have reviewed the documentation and source code for this token, checking for unexpected behaviours that could affect the pool.
      Note: Openzeppelin libraries used for a majority of the code base
      Audit by Peckshield: <https://github.com/stafiprotocol/stafi-bootstrap/blob/master/audits/reth/PeckShield-Audit-Report-StaFi-Eth2-Staking-v1.0.pdf>

### Incenvites

This pool WILL receive incentives other than CRV.

### Fees

1. Does this pool include tokens that are already handled by the fee burning process? No

2. Does this pool include tokens that are not currently being burned, but which can be handled by an already deployed burner? 
Yes, the synthBurner will suffice

3. Does this pool require deployment of any new burners? Shouldn't be necessary

### Pre-review Checklist

- [x] I have included a `pooldata.json` with accurate token addresses
- [x] I have included a `README.md` with accurate token addresses
- [x] I have added a workflow targeting this pool, and confirmed that all tests are passing within the CI
- [x] I have run `brownie test tests/forked --network mainnet-fork` locally and confirmed that all tests are passing
- [ ] I have verified the fee burning process for this pool and opened a related pull request on [`curve-dao-contracts`](https://github.com/curvefi/curve-dao-contracts) if needed
   Not verified but PR has been submitted curvefi/curve-dao-contracts#95

### Reviewer Pre-deployment Checklist

- [ ] This PR contains a new workflow file, and the entire test suite is passing
- [ ] I have verified the success / failure behaviours for each token in this pool
- [ ] I have run `brownie test tests/forked --network mainnet-fork` locally and confirmed that all tests are passing
- [ ] I have verified the fee burning process for this pool and any related pull requests
- [ ] This pool makes no significant modifications to the core math OR has been reviewed personally by [@michwill](https://github.com/michwill)

### Post-deployment Checklist

- [ ] I have deployed a [`LiquidityGaugeV2`](https://github.com/curvefi/curve-dao-contracts/blob/master/contracts/gauges/LiquidityGaugeV2.vy) contract for this pool
- [ ] I have deployed any required fee burner contracts
- [ ] I have added pool and gauge deployment addresses to the `pooldata.json` and `README.md` files within this PR
- [ ] I have opened a pull request on [`curve-docs`](https://github.com/curvefi/curve-docs) to add the new deployment addresses to the documentation
- [ ] I have [created a DAO vote]((https://github.com/curvefi/curve-dao-contracts/blob/master/scripts/voting/new_vote.py)) to add the new gauge and enable/configure fee burning

### CODE REVIEW

**Comment by skellet0r** [comment]
No big brain stuff, mainly just a ctrl+c and ctrl+v of the ankrETH pool, with minor edits.

File: tests/conftest.py [hunk 1]
```diff
@@ -103,7 +104,7 @@ def pytest_ignore_collect(path, config):
         return None
 
     # always allow forked tests
-    if path_parts[0] == "forked":
+    if path_parts[:1] == "forked":
```
Comment (iamdefinitelyahuman) [inline tests/conftest.py:107]:
```suggestion
    if path_parts[:1] == ("forked",):
```

File: contracts/pools/reth/StableSwapRETH.vy [hunk 2]
```diff
@@ -0,0 +1,843 @@
...
+@internal
+def _xp(rates: uint256[N_COINS]) -> uint256[N_COINS]:
+    result: uint256[N_COINS] = rates
+    for i in range(N_COINS):
+        result[i] = result[i] * self.balances[i] / PRECISION
+    return result
+
+
+@internal
+@view
```
Comment (michwill) [inline contracts/pools/reth/StableSwapRETH.vy:208]: Although it doesn't matter at all, but can as well be `@pure`

**Review by michwill — APPROVED** [review]
Looks good to me

**Comment by skellet0r** [comment]
Note to self: Add rETH rate method ID in the curve-dao-contracts add_pool script.

**Comment by charlie-eth** [comment]
@skellet0r The pool will receive incentives other than CRV

File: contracts/pools/reth/StableSwapRETH.vy [hunk 3]
```diff
@@ -0,0 +1,843 @@
...
+
+@view
+@external
+def get_virtual_price() -> uint256:
+    """
+    @notice The current virtual price of the pool LP token
+    @dev Useful for calculating profits
+    @return LP token virtual price normalized to 1e18
+    """
+    D: uint256 = self.get_D(self._xp(self._stored_rates()), self._A())
```
[the remark this answers is not in the document]
Comment (skellet0r) [reply contracts/pools/reth/StableSwapRETH.vy:255]: Just to be safe as soon as Vyper 0.2.12 is released and vyperlang/vyper#2345 is included, I'll update the version pragma on this

**Comment by skellet0r** [comment]
Uh oh, looks like I'll have to submit a seperate PR for the templates,  v0.2.12 is causing some compilation errors in the templates.

**Review by iamdefinitelyahuman — APPROVED** [review]

**Pull request merged**